src/util/trace.h

```diff
@@ -7,39 +7,58 @@
+// used. Here, an expensive argument preparation could, for example, be
+// calculating a hash or serialization of a data structure.
+#define _SDT_HAS_SEMAPHORES 1
+
 #include <sys/sdt.h>
 
-#define TRACE(context, event) DTRACE_PROBE(context, event)
-#define TRACE1(context, event, a) DTRACE_PROBE1(context, event, a)
-#define TRACE2(context, event, a, b) DTRACE_PROBE2(context, event, a, b)
-#define TRACE3(context, event, a, b, c) DTRACE_PROBE3(context, event, a, b, c)
-#define TRACE4(context, event, a, b, c, d) DTRACE_PROBE4(context, event, a, b, c, d)
-#define TRACE5(context, event, a, b, c, d, e) DTRACE_PROBE5(context, event, a, b, c, d, e)
-#define TRACE6(context, event, a, b, c, d, e, f) DTRACE_PROBE6(context, event, a, b, c, d, e, f)
-#define TRACE7(context, event, a, b, c, d, e, f, g) DTRACE_PROBE7(context, event, a, b, c, d, e, f, g)
-#define TRACE8(context, event, a, b, c, d, e, f, g, h) DTRACE_PROBE8(context, event, a, b, c, d, e, f, g, h)
-#define TRACE9(context, event, a, b, c, d, e, f, g, h, i) DTRACE_PROBE9(context, event, a, b, c, d, e, f, g, h, i)
-#define TRACE10(context, event, a, b, c, d, e, f, g, h, i, j) DTRACE_PROBE10(context, event, a, b, c, d, e, f, g, h, i, j)
-#define TRACE11(context, event, a, b, c, d, e, f, g, h, i, j, k) DTRACE_PROBE11(context, event, a, b, c, d, e, f, g, h, i, j, k)
-#define TRACE12(context, event, a, b, c, d, e, f, g, h, i, j, k, l) DTRACE_PROBE12(context, event, a, b, c, d, e, f, g, h, i, j, k, l)
+// Used to define a counting semaphore for a tracepoint. This semaphore is
+// automatically incremented by tracing frameworks (bpftrace, bcc, libbpf, ...)
+// upon attaching to the tracepoint and decremented when detaching. This needs
+// to be a global variable. It's placed in the '.probes' ELF section.
+#define TRACEPOINT_SEMAPHORE(context, event) \
+    unsigned short context##_##event##_semaphore __attribute__((section (".probes")))
+
+// Returns true if something is attached to the tracepoint.
+#define TRACEPOINT_ACTIVE(context, event) context##_##event##_semaphore > 0
+
+// A USDT tracepoint with zero to twelve arguments. Here, it **IS** checked that
+// the tracepoint is active before preparing its arguments. Use the
+// TRACEPOINT_UNCHECKED macro if you want to manually check that the tracepoint
+// is active.
+#define TRACEPOINT(context, event, args...) \
+    if (TRACEPOINT_ACTIVE(context, event)) \
+        STAP_PROBEV(context, event, args)
+
+// A USDT tracepoint with zero to twelve arguments. Here, it is **NOT** checked
+// that the tracepoint is active before preparing its arguments. Use this macro
+// if you want to, for example, prepare expensive tracepoint arguments over
+// multiple lines. Example usage:
+//
+//  if (TRACEPOINT_ACTIVE(context, event)) {
+//      result = slightly_expensive_calulation();
+//      TRACEPOINT_UNCHECKED(context, event, result);
+//  }
+//
+#define TRACEPOINT_UNCHECKED(context, event, args...) \
```

Comment:
Aren't variadic macros (at least in this form) an extension, not standard C++?

Reply:
TIL, could change to `...` and `__VA_ARGS__` if needed